crates/rattler_networking/src/s3_middleware.rs

```diff
@@ -0,0 +1,468 @@
+                }
+                (_, Some(_)) => {
+                    return Err(anyhow::anyhow!("Unsupported authentication method"));
+                }
+                (_, None) => {
+                    tracing::info!("No authentication found, assuming bucket is public");
+                    let sdk_config = aws_config::defaults(BehaviorVersion::latest())
+                        .no_credentials() // Turn off request signing
+                        .load()
+                        .await;
+                    aws_sdk_s3::config::Builder::from(&sdk_config)
+                        .endpoint_url(endpoint_url)
+                        .region(aws_sdk_s3::config::Region::new(region))
+                        .force_path_style(force_path_style)
+                }
+            };
+            let s3_config = config_builder.build();
+            Ok(aws_sdk_s3::Client::from_conf(s3_config))
+        } else {
+            let mut sdk_config = aws_config::defaults(BehaviorVersion::latest()).load().await;
+            if let Some(credentials_provider) = sdk_config.credentials_provider() {
+                let creds = credentials_provider.as_ref().provide_credentials().await;
+                if creds.is_ok() {
+                    tracing::info!("Using AWS credentials from environment via default provider");
+                } else {
+                    tracing::warn!("No AWS credentials found, assuming bucket is public");
+                    sdk_config = aws_config::defaults(BehaviorVersion::latest())
+                        .no_credentials()
+                        .load()
+                        .await;
+                }
+            }
+            let mut s3_config_builder = aws_sdk_s3::config::Builder::from(&sdk_config);
+            // Infer if we expect path-style addressing from the endpoint URL.
+            if let Some(endpoint_url) = sdk_config.endpoint_url() {
+                // If the endpoint URL is localhost, we probably have to use path-style addressing.
+                // There are certainly more edge cases, but this is a valid start to make the
+                // integration tests with minIO work.
+                // xref: https://github.com/awslabs/aws-sdk-rust/issues/1230
+                if endpoint_url.starts_with("http://localhost") {
+                    s3_config_builder = s3_config_builder.force_path_style(true);
+                }
+            }
+            let client = aws_sdk_s3::Client::from_conf(s3_config_builder.build());
+            Ok(client)
+        }
+    }
+
+    /// Generate a presigned S3 `GetObject` request.
+    async fn generate_presigned_s3_url(&self, url: Url) -> MiddlewareResult<Url> {
+        let client = self.create_s3_client(Some(url.clone())).await?;
+
+        let bucket_name = url.host_str().ok_or_else(|| {
+            reqwest_middleware::Error::middleware(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "Host should be present in S3 URL",
```

Comment:
please start library errors with lowercase. That makes it easier to concatenate errors like: "Something when wrong because, {}".

```suggestion
                "host is missing from s3:// URL",
```

Reply:
Done!